Add access to dashboard

mook-as opened the pull request:
This adds a menu item to open the rancher dashboard (i.e. `cattle-system` service `homestead`).  This sets up a port forwarding via the main node process.

Note that while the dashboard loads, it appears to be rather non-functional at the moment.

Fixes #8.

mook-as commented on background.js:
```diff
@@ -171,6 +171,15 @@ ipcMain.on('k8s-restart', async (event) => {
   }
 });
 
+app.on('window-preferences', () => {
```
I didn't love tacking random events onto the app; I think a better fit would be to convert the tray stuff to a class that extends `EventEmitter`, and have that emit commands instead.  But that seemed a bit big for that PR — I'd be happy to do a follow-up to implement that (which would be slightly easier to review than doing it in this one).

mattfarina replied:
SGTM. It can be here or a follow-up. Your call.

mook-as replied:
I'd prefer in a follow-up, just to keep the PRs small and focused(-ish).

mattfarina reviewed (commented):
While I have not had a chance to completely review the PR.... the change is not working. The dashboard does not open for me. In the console output I get

```
(node:77445) [DEP0123] DeprecationWarning: Setting the TLS ServerName to an IP address is not permitted by RFC 6066. This will be ignored in a future version.
(node:77445) UnhandledPromiseRejectionWarning: #<ErrorEvent>
(node:77445) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:77445) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

The window opens for the dashboard but contains no content.

mook-as commented:
I believe the issue is that the homestead pod takes a while to come up.  Added some code to wait for the endpoint to be alive before allowing the dashboard to be accessed.  This also includes adding a new state (`READY`), since that could take a while.

mattfarina commented:
The dashboard is not working but that is not related to this PR. I currently see ....

<img width="1052" alt="Screen Shot 2021-01-12 at 11 22 18 AM" src="https://user-images.githubusercontent.com/62991/104343441-f1fd4880-54c9-11eb-9256-a225a6a0b7d5.png">

This looks to be related to something with homestead (rancher) reaching out to an endpoint that's not working.

mattfarina commented on src/k8s-engine/client.js:
```diff
@@ -0,0 +1,147 @@
+            /** @type k8s.V1EndpointsList */
+            let endpoints;
+            ({ body: endpoints } = await this.#coreV1API.listNamespacedEndpoints(namespace, { headers: { name: endpointName } }));
+            console.log(`Got ${endpointName} endpoints: ${endpoints}`);
+            target = endpoints?.items?.pop()?.subsets?.pop()?.addresses?.pop()?.targetRef;
+            console.log(`Got ${endpointName} target: ${target?.namespace}:${target?.name}`);
```
These are causing repeated log entries of ...

```
Got homestead endpoints: [object Object]
Got homestead target: undefined:undefined
```

The `[object Object]` should be fixed. At what point does this not get written to the log on all the loops? Should we add a debug mode for that?

mook-as commented:
Reduced the logging (at leas, stopped dumping giant error objects to stdout).

Not sure why I needed that last commit — 613bdb82c5de204a8109d308819ed81e2d9f7956 didn't even modify the affected file. But it should be ready for re-review now, @mattfarina.

mattfarina reviewed (changes requested):
I've not figured out where in the code this is happening but the "Reset Kubernetes" button no longer works in this PR. You click it, the button goes grey, but nothing happens. If you close and open prefs again it's ready to be clicked.

mook-as commented:
The reset kubernetes function in the main process had a reversed check; this is now fixed.

I also changed things a bit so that we can mark the client as shutting down, so that when you do a reset it can stop trying to give you a dashboard port forward.

mattfarina reviewed (changes requested):
The dashboard seems to be a hard one because of all the impacts.

I opened the dashboard and then chose to switch my Kubernetes version. This causes k8s to shut down, delete the instance, and start a fresh setup. When I did that an alert error popped up with:

```
Uncaught Exception:
Error: WebSocket is not open: readyState 3 (CLOSED)
    at WebSocket.send (webpack:///./node_modules/ws/lib/websocket.js?:330:19)
    at Socket.eval (webpack:///./node_modules/@kubernetes/client-node/dist/web-socket-handler.js?:165:12)
    at Socket.emit (events.js:310:20)
    at addChunk (_stream_readable.js:286:12)
    at readableAddChunk (_stream_readable.js:268:9)
    at Socket.Readable.push (_stream_readable.js:209:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:186:23)
```

This needs to be fixed.

mook-as commented:
- Rebased.
- Added a custom wrapper around the socket for the port forwarding, so that we could drop any errors from port forwarding on the floor.  Unfortunately the port forwarding API doesn't have support for doing this in a non-ugly way, as far as I can tell.  We also currently don't have anywhere sensible to report this, so it's just console for now.
- Changed things to automatically close the dashboard window when the dashboard is not ready.  I'm doing this because if the user clicks around while the dashboard is down, they're likely to end up in a state where they end up at a blank screen and the only way to get back out is to close the window and re-open it anyway.
